openupgradelib/migrator.py

```diff
@@ -0,0 +1,549 @@
+# -*- coding: utf-8 -*-
+# © 2016 Grupo ESOC Ingeniería de Servicios, S.L.U. - Jairo Llopis
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+import logging
+from contextlib import contextmanager
+from functools import wraps
+from psycopg2 import errorcodes, ProgrammingError
+from openerp import api, SUPERUSER_ID
+from .openupgrade import allow_pgcodes
+
+_logger = logging.getLogger(__name__)
+
+
+# TODO Remove when module ends WIP stage
+def wip(function):
+    """This indicates methods that are currently WIP."""
+    @wraps(function)
+    def wrapper(*args, **kwargs):
+        _logger.warning(
+            "BEWARE! You are using an untested function: %s. If you find it "
+            "works fine, please remove the `@wip` decorator.",
+            function.__name__)
+        return function(*args, **kwargs)
+    return wrapper
+
+
+class MigrationException(Exception):
+    """Custom exception raised only by the migration toolkit.
+
+    Odoo handles some exceptions like :class:`AttributeError` and
+    :class:`psycopg2.OperationalError` in some places of the upper stack. By
+    defining a custom class for this toolkit's exceptions, you make Odoo not
+    handle them and abort the migration process with a useful error message.
+    """
+    def __init__(self, old):
+        """Create exception and link it to the older one.
+
+        :param Exception old:
+            Old exception that was caught before this one.
+        """
+        _logger.exception(old)
+        self.old = old
+        super(MigrationException, self).__init__(*old.args)
+
+
+def manage(function):
+    """Run sentences inside a transaction and environment.
+
+    If anything fails inside :param:`function`, all will be rolled back; and
+    you have access to an Odoo environment like when using models normally.
+
+    This decorator is prepared to be applied to your ``migrate`` method (the
+    required method for all migrations), but if you want to use it in another
+    context, at least methods decorated with this must have ``cr`` as their
+    first parameter.
+    """
+    @wraps(function)
+    def wrapper(cr, *args, **kwargs):
+        try:
+            with cr.savepoint():
+                with api.Environment.manage():
```

Comment:
@Yajo How would usage of this look like? How would I include that into the current migrate decorator? I think this would be rather easy to include, wouldn't it?